modules/FvwmTheme/FvwmTheme.1
.\" t
.\" @(#)FvwmTheme.1  7/20/1999
.TH FvwmTheme 1 "3 July 2001"
.UC

.SH NAME
FvwmTheme \- an FVWM modules for managing the appearance of FVWM and its modules

.SH SYNOPSIS
FvwmTheme is spawned by fvwm, so no command line invocation will work.

.SH COPYRIGHT
Copyright (C) 1999 Joey Shutup.  No guarantees or warranties or anything are
provided or implied in any way whatsoever.  Use this program at your own risk.
Permission to use this program for any purpose is given, as long as the
copyright is kept intact.

.SH DESCRIPTION
FvwmTheme creates appearance resources that may be shared by FVWM and other
modules.  It reads an initial configuration and also reacts to configuration
commands and messages sent from FVWM so that the resources can be dynamically
changed.

.SH INVOCATION
FvwmTheme must be spawned as a module by fvwm.
.PP
It is important that FvwmTheme is invoked before any other modules
that use the color sets provided by FvwmTheme.  Thus FvwmTheme has
to be invoked with the 'ModuleSynchronous' command by inserting the
line 'ModuleSynchronous Timeout 5 FvwmTheme' in the .fvwm2rc file
right after the \fIModulePath\fP and \fIImagePath\fP have been defined.
Invoking FvwmTheme from the \fIInitFunction\fP, \fIStartFunction\fP
or\fIRestartFunction\fP or later may not work properly and can
cause excessive redrawing of already running modules.  It is important
that the configuration lines for FvwmTheme appear in the configuration
file before FvwmTheme is started.  You can find a proper sample
fvwm setup at the end of this document.  Fvwm will search the
directory specified in the ModulePath configuration option to attempt
to locate FvwmTheme.  It is pointless to run more than one FvwmTheme
so there is no provision for using an alias name.

.SH CONFIGURATION OPTIONS
FvwmTheme supports some configuration options.

.IP "*FvwmTheme: Colorset \fIn\fP \fIoptions\fP "
Will create or modify colorset \fIn\fP. Each colorset has four colors, an
optional pixmap and an optional shape mask.  The four colors are used by
modules as the foreground, background, highlight and shadow colors.  When
a colorset is created it defaults to a foreground of black and background of
gray.  The background and foreground are marked as "average" and "contrast"
(see later) so that just specifying a pixmap or gradient gives sensible
results.

Warning: The highest colorset number used determines memory consumption.
Thus if you define 'Colorset 100000', the memory for 100001 colorsets will
be used.  Keep your colorset numbers as small as possible.

\fIoptions\fP is a comma separated list containing some of the keywords:
fg, Fore, Foreground, bg, Back, Background, hi, Hilite, Hilight, sh,
Shade, Shadow, Pixmap, TiledPixmap, AspectPixmap, Shape, TiledShape,
AspectShape, NoShape, ?Gradient, Plain and Transparent.

\fIfg\fP, \fIFore\fP and \fIForeground\fP take a color name as an argument and
set the foreground color.  The special name \fIContrast\fP may be used to
select a color that contrasts well with the background color.  To reset
the foreground color to the default value you can simply omit the color
name.

\fIbg\fP, \fIBack\fP and \fIBackground\fP take a color name as an argument and
set the background color.  It also sets the highlight and shadow colors to
values that give a 3d effect unless these have been explicitly set with the
options below.  The special name \fIAverage\fP may be used to select a color
that is the average color of the pixmap.  The background color is reset to
the default value if the color name is omitted.

\fIhi\fP, \fIHilite\fP and \fIHilight\fP take a color name as as argument and
set the highlight color.  If the hilight color is not explicitly set,
the default is to calculate it from the background color.  To switch back
to the default behavior the color name can be omitted.

\fIsh\fP, \fIShade\fP and \fIShadow\fP take a color name as as argument and
set the shadow color.  If the shadow color is not explicitly set,
the default is to calculate it from the background color.  To switch back
to the default behavior the color name can be omitted.

\fIPixmap\fP, \fITiledPixmap\fP and \fIAspectPixmap\fP take a file name as
an argument, search the \fIImagePath\fP and use it as the background pixmap.
Any transparent parts are filled with the background color.  Not specifying a
file name removes any existing one from the colorset.  \fITiledPixmap\fP will
produce repeated copies of the image with no scaling, \fIPixmap\fP will
cause the image to be stretched to fit whatever object the colorset is applied
to and \fIAscpectPixmap\fP will stretch to fit but will retain the image
aspect ratio.  Bitmaps are not supported.

\fITransparent\fP tries to create a transparent background pixmap.
It may be necessary to reapply this option whenever the background
of one of the affected application changes.
The pixmap may be used as a window background to achieve root transparency.
The root background change may be detected or not, this depends on the program
used to set the background. If you use xpmroot, xsetbg (xli),
FvwmBacker with a solid or a colorset colors or a recent version of Esetroot
(>= 9.2) a background change should be detected. If background changes are not
detected (e.g., if you use xv or xsetroot) you can force detection by using
the -d option of xpmroot:
.in +.3i
xv -root -quit mybg.png; xpmroot -d
.in -.3i
Due to the way X implements transparency no guarantees can be made
that the desired effect can be achieved.  The application may even crash.
If you experience any problems with this option, do not use it.

\fIShape\fP, \fITiledShape\fP and \fIAspectShape\fP take a file name as
an argument, search the \fIImagePath\fP and use it as the shape bitmap.
\fITiledShape\fP will produce repeated copies of the bitmap with no scaling,
\fIShape\fP will cause the bitmap to be stretched to fit whatever object the
colorset is applied to and \fIAscpectShape\fP will stretch to fit but will
retain the bitmap aspect ratio.  If the file is a pixmap in xpm format,
the shape mask of the pixmap is used.

Warning: Due to the way X11 implements shapes and the implementation
of the FvwmTheme module you cannot take back making windows shaped.
You may have to restart fvwm or the shaped application.

\fI?Gradient ...\fP will create a pixmap and stretch it to fit the window.
\fI?Gradient\fP may be one of Hgradient, Vgradient, Dgradient, Bgradient,
Sgradient, Cgradient, Rgradient or Ygradient.  The gradient types are as
follows:  H is horizontal; V is vertical; D is diagonal from top left to
bottom right; B is a backwards diagonal from bottom left to top right; S
is concentric squares; C is con-centric circles; R is a radar like pattern
and Y is a Yin Yang style (but without the dots, we are not \fIthat\fP mad).
Please refer to the \fICOLOR GRADIENTS\fP section in the \fIfvwm2\fP man page
for the syntax of gradients.

\fINoShape\fP removes tha shape mask from the colorset while \fIPlain\fP
removes the background pixmap or gradient.

.IP "*FvwmTheme: ReadWriteColors"
Tells FvwmTheme to use read/write color cells for the foreground and background
pixels.  This can make changing colorsets much faster but can only be used with
PseudoColor displays.  This option must be specified first and it cannot be
turned off.  If the pixmap and shape of a colorset are unchanged modules
do not redraw when the colors are changed.

.SH COMMANDS
The following FVWM command may be executed at any time to alter the colorsets.
It may be bound to a menu item or typed into a module such as FvwmConsole.

.IP "SendToModule FvwmTheme Colorset \fIoptions\fP"
The syntax is the same as the configuration option.

.SH EXAMPLES

.in +.3i
*FvwmTheme: Colorset 3 fg wheat, bg navy
.in -.3i

If necessary this will create colorsets 0, 1, 2 and 3 and then change colorset
3 to have a foreground of wheat, a background of navy.

.in +.3i
*FvwmTheme: Colorset 3 bg "navy blue"
.in -.3i

will change the background color of colorset 3 to navy blue. The foreground and
pixmap will be unchanged.

.in +.3i
*FvwmTheme: Colorset 3 AspectPixmap \\
  large_murky_dungeon.xpm
.in -.3i

will cause depression

.in +.3i
*FvwmTheme: Colorset 3 bg Average
.in -.3i

will set the background color and the relief colors to match the background
pixmap. This is the default setting but it must be used if a background color
was specified and is now not required.

.in +.3i
*FvwmTheme: Colorset 3 YGradient 200 3 \\
  blue 1000 navy 1 blue 1000 navy
.in -.3i

will add a Yin Yang gradient background pixmap to colorset 3.  If the
background is set to average it will be recomputed as will the foreground
if that is set to contrast.

.EX
 #!/bin/sh
 FvwmCommand "SendToModule FvwmTheme Colorset 7
   fg navy, bg gray"
 while true
 do
   FvwmCommand "SendToModule FvwmTheme Colorset 7
     fg gray"
   sleep 1
   FvwmCommand "SendToModule FvwmTheme Colorset 7
     fg navy"
   sleep 1
 done
.EE

will make colorset 7 blink if used with the *FvwmTheme: ReadWriteColors option.

The color names used in color sets can be substituted in any fvwm
command.  Please refer to the \fICOMMAND EXPANSION\fP
section in the fvwm2 man page and the example below for a description.

.SH SAMPLE FVWM CONFIGURATION

Below you can find a fvwm configuration file that demonstrates the
use of the FvwmTheme module.  The order in which FvwmTheme and the
other modules are configured and started is important.

.EX
 # where your images are
 ImagePath <put your image path here>

 #
 # FvwmTheme
 #
 # The FvwmTheme setup must be first in the config file,
 # right after the paths are set.
 #
 # Instead of the *FvwmTheme: Colorset... lines below you
 # could read in a file with these commands.  So to change
 # your color scheme you can simply copy a different file
 # over your palette file and restart fvwm:
 #
 # Read /home/my_user_name/.fvwm2_palette
 #

 # 0 = Default colours
 # 1 = Inactive windows
 # 2 = Active windows
 # 3 = Inactive menu entry and menu background
 # 4 = Active menu entry
 # 5 = greyed out menu entry (only bg used)
 # 6 = module foreground and background
 # 7 = hilight colours
 *FvwmTheme: Colorset 0 fg black, bg rgb:b4/aa/94
 *FvwmTheme: Colorset 1 fg black, bg rgb:a1/b2/c8
 *FvwmTheme: Colorset 2 fg black, bg rgb:da/9a/68
 *FvwmTheme: Colorset 3 fg black, bg rgb:b4/aa/94, \\
   VGradient 100 dtcolour5 rgb:b4/aa/94
 *FvwmTheme: Colorset 4 fg black, bg rgb:b4/aa/94
 *FvwmTheme: Colorset 5 fg rgb:d2/bf/a8, \\
   bg rgb:b4/aa/94
 *FvwmTheme: Colorset 6 fg black, bg rgb:b4/aa/94, \\
   VGradient 100 dtcolour5 rgb:b4/aa/94
 *FvwmTheme: Colorset 7 fg black, bg rgb:94/ab/bf

 # run FvwmTheme before anything else is done
 ModuleSynchronous Timeout 5 FvwmTheme

 #
 # general setup
 #
 Style * Colorset 1
 Style * HilightColorset 2
 MenuStyle * MenuColorset 3
 MenuStyle * ActiveColorset 4
 MenuStyle * GreyedColorset 5

 #
 # Applications
 #
 AddToFunc InitFunction
 + I Exec exec xterm -fg $[fg.cs0] -bg $[bg.cs0]

 #
 # module setup
 #

 # ... more FvwmPager config lines ...
 *FvwmPager: Colorset * 6
 *FvwmPager: BalloonColorset * 6
 *FvwmPager: HilightColorset * 7
 *FvwmPager: WindowColorsets 1 2

 # ... more FvwmIconMan config lines ...
 *FvwmIconMan: *Colorset 6
 *FvwmIconMan: *FocusColorset 2
 *FvwmIconMan: *FocusAndSelectColorset 2
 *FvwmIconMan: *PlainColorset 6
 *FvwmIconMan: *SelectColorset 6
 *FvwmIconMan: *TitleColorset 6

 # ... more FvwmButtons config lines ...
 *FvwmButtons: Colorset 6
 # sample button passing color to xterm
 *FvwmButtons: (Title xterm, \\
   Action "Exec exec xterm -fg $[fg.cs6] -bg[bg.cs6]")

 # ... more FvwmWharf config lines ...
 *FvwmWharf: Colorset 6

 # ... more FvwmIdent config lines ...
 *FvwmIdent: Colorset 6

 # ... more FvwmWinList config lines ...
 *FvwmWinList: Colorset      1
 *FvwmWinList: FocusColorset 2
 *FvwmWinList: IconColorset  1

 # ... more FvwmTaskBar config lines ...
 *FvwmTaskBar: Colorset     6
 *FvwmTaskBar: IconColorset 6
 *FvwmTaskBar: TipsColorset 0
.EE

If you need to have more colors and don't want to reinvent the wheel,
you may use the convention used in fvwm-themes, it defines the meaning
of the first 40 colorsets for nearly all purposes:

 \fBhttp://fvwm-themes.sourceforge.net/doc/colorsets\fP

.SH BUGS

Initialization of fvwm, FvwmTheme and the other modules is tricky.
Please pay close attention to the text in the \fIINVOCATION\fP
section.  The example above demonstrates the proper way to get a
FvwmTheme setup running.

.SH AUTHOR

Prefers to remain anonymous.
With help from Brad Giaccio and Dominik Vogt.


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmTheme/FvwmTheme.c
/* Copyright (C) 1999 Joey Shutup.
 *
 * http://www.streetmap.co.uk/streetmap.dll?postcode2map?BS24+9TZ
 *
 * No guarantees or warranties or anything are provided or implied in any way
 * whatsoever.  Use this program at your own risk.  Permission to use this
 * program for any purpose is given, as long as the copyright is kept intact.
 */
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "config.h"

#include <stdio.h>
#include <signal.h>

#include <X11/Xlib.h>
#include <X11/Xproto.h>
#include <X11/Intrinsic.h>

#ifdef SHAPE
#include <X11/extensions/shape.h>
#endif /* SHAPE */

#include "libs/fvwmlib.h"
#include "libs/Module.h"
#include "libs/Picture.h"
#include "libs/Colorset.h"

/* Globals */
static Display *dpy;
static Window win;	/* need a window to create pixmaps */
static GC gc;		/* ditto */
static char *name;
static int color_limit = 0;
static int fd[2];	/* communication pipes */

/* forward declarations */
static RETSIGTYPE signal_handler(int signal);
static void set_signals(void);
static int error_handler(Display *dpy, XErrorEvent *e);
static void parse_config(void);
static void parse_config_line(char *line);
static void parse_message_line(char *line);
static void main_loop(void) __attribute__((__noreturn__));
static void parse_colorset(char *line);

int main(int argc, char **argv)
{
  XSetWindowAttributes xswa;
  XGCValues xgcv;

  name = GetFileNameFromPath(argv[0]);

  set_signals();

  /* try to make sure it is fvwm that spawned this module */
  if (argc != 6) {
    fprintf(stderr, "%s "VERSION" should only be executed by fvwm!\n", name);
    exit(1);
  }

  /* note the communication pipes */
  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  /* open the display and initialize the picture library */
  dpy = XOpenDisplay(NULL);
  if (!dpy) {
    fprintf(stderr,"%s: can't open display %s", name, XDisplayName(NULL));
    exit (1);
  }
  InitPictureCMap(dpy);

  /* This module allocates resouces that othe rmodules may rely on.
   * Set the closedown mode so that the pixmaps and colors are not freed
   * by the server if this module dies
   */
  XSetCloseDownMode(dpy, RetainTemporary);

  /* create a window to work in */
  xswa.background_pixmap = CopyFromParent;
  xswa.border_pixel = 0;
  xswa.colormap = Pcmap;
  win = XCreateWindow(dpy, RootWindow(dpy, DefaultScreen(dpy)), -2, -2, 2, 2, 0,
		      Pdepth, InputOutput, Pvisual,
		      CWBackPixmap | CWBorderPixel | CWColormap, &xswa);

  /* create a GC */
  gc = XCreateGC(dpy, win, 0, &xgcv);

  /* die horribly if either win or gc could not be created */
  XSync(dpy, False);

  /* install a non-fatal error handler */
  XSetErrorHandler(error_handler);

  /* get the initial configuration options */
  parse_config();

  /* tell fvwm we're running */
  SendFinishedStartupNotification(fd);

  /* garbage collect */
  alloca(0);

  /* just in case any previous FvwmTheme has died and left pixmaps dangling.
   * This might be overkill but any other method must ensure that fvwm doesn't
   * get killed (it can be the owner of the pixels in colorset 0)
   */
  XKillClient(dpy, AllTemporary);

  /* sit around waiting for something to do */
  main_loop();
}

static void main_loop(void)
{
  FvwmPacket *packet;

  while (True) {
    /* garbage collect */
    alloca(0);

    /* wait for an instruction from fvwm */
    packet = ReadFvwmPacket(fd[1]);
    if (!packet)
      exit(0);

    /* handle dynamic config lines and text messages */
    switch (packet->type) {
    case M_CONFIG_INFO:
      parse_config_line((char *)&packet->body[3]);
      break;
    case M_STRING:
      parse_message_line((char *)&packet->body[3]);
      break;
    }
  }
}

/* config options, the first NULL is replaced with *FvwmThemeColorset */
/* the "Colorset" config option is what gets reflected by fvwm */
static char *config_options[] =
{
  "ImagePath",
  "ColorLimit",
  "Colorset",
  NULL,
  NULL
};

static void parse_config_line(char *line)
{
  char *rest;

  switch(GetTokenIndex(line, config_options, -1, &rest)) {
  case 0:
    SetImagePath(rest);
    break;
  case 1:
    sscanf(rest, "%d", &color_limit);
    break;
  case 2:
    /* got a colorset config line from fvwm, maybe Defaultcolors: load it */
    LoadColorsetAndFree(rest);
    break;
  case 3:
    parse_colorset(rest);
    break;
  }
}

/* background pixmap parsing */
static char *pixmap_options[] =
{
  "TiledPixmap",
  "Pixmap",
  "AspectPixmap",
  NULL
};
static char *shape_options[] =
{
  "Shape",
  "TiledShape",
  "AspectShape",
  NULL
};

static char *dash = "-";
static char *average = "Average";

/* translate a colorset spec into a colorset structure */
static void parse_colorset(char *line)
{
  int n, ret;
  int i;
  colorset_struct *cs;
  char *token, *fg, *bg;
  Picture *picture;

  /* find out which colorset and make sure it exists */
  token = PeekToken(line, &line);
  if (!token) {
    return;
  }
  ret = !sscanf(token, "%d", &n);
  if (ret)
    return;
  cs = AllocColorset(n);

  /* get the mandatory foreground and background color names */
  line = GetNextToken(line, &fg);
  line = GetNextToken(line, &bg);
  if (!fg)
    return;
  if (!bg) {
    free(fg);
    return;
  }

  /* fill in the structure if color name is specified */
  if (!StrEquals(fg, dash)) {
    XFreeColors(dpy, Pcmap, &cs->fg, 1, 0);
    cs->fg = GetColor(fg);
  }
  if (!StrEquals(bg, dash) && !StrEquals(bg, average)) {
    XFreeColors(dpy, Pcmap, &cs->bg, 3, 0);
    cs->bg = GetColor(bg);
    cs->hilite = GetHilite(cs->bg);
    cs->shadow = GetShadow(cs->bg);
  }

  /* look for a pixmap specifier, if not "-" remove the existing one */
  token = PeekToken(line, &line);
  if (token) {
    ret = StrEquals(token, dash);
  }
  if (!ret) {
    XFreePixmap(dpy, cs->pixmap);
    XFreePixmap(dpy, cs->mask);
    XFreePixmap(dpy, cs->shape_mask);
    cs->pixmap = None;
    cs->mask = None;
    cs->shape_mask = None;
  }

  i = GetTokenIndex(token, shape_options, 0, NULL);
  if (i != -1)
  {
    /* read filename */
    token = PeekToken(line, &line);
#ifdef SHAPE
    /* set the flags */
    cs->shape_tile = 0;
    cs->shape_keep_aspect = 0;
    if (i == 1)
      cs->shape_tile = 1;
    else if (i == 2)
      cs->shape_keep_aspect = 1;

    /* try to load the shape mask */
    if (token && !ret)
    {
      /* load the shape mask */
      picture = GetPicture(dpy, win, NULL, token, color_limit);
      if (!picture)
	fprintf(stderr, "%s: can't load picture %s\n", name, token);
      else if (picture->depth != 1 && picture->mask == None)
      {
	fprintf(stderr, "%s: shape pixmap must be of depth 1\n", name);
	DestroyPicture(dpy, picture);
      }
      else
      {
	Pixmap mask;

	/* okay, we have what we want */
	if (picture->mask != None)
	  mask = picture->mask;
	else
	  mask = picture->picture;

	if (mask != None)
	{
	  cs->shape_mask = XCreatePixmap(
	    dpy, mask, picture->width, picture->height, 1);
	  if (cs->shape_mask != None)
	  {
	    XGCValues xgcv;
	    static GC shape_gc = None;

	    if (shape_gc == None)
	    {
	      xgcv.foreground = 1;
	      xgcv.background = 0;
	      /* create a gc for 1 bit depth */
	      shape_gc = XCreateGC(
		dpy, picture->mask, GCForeground|GCBackground, &xgcv);
	    }
	    XCopyPlane(dpy, mask, cs->shape_mask, shape_gc, 0, 0,
		       picture->width, picture->height, 0, 0, 1);
	  }
	}
      }
      if (picture)
      {
	DestroyPicture(dpy, picture);
	picture = None;
      }
    }
#else
    cs->shape_mask = None;
#endif
    /* skip filename */
    token = PeekToken(line, &line);
  }

  /* if one was specified try to load it */
  if (token && !ret)
  {
    int type = GetTokenIndex(token, pixmap_options, 0, NULL);
    unsigned int w, h;

    switch(type)
    {
    case 0:
    case 1:
    case 2:
      /* all three of these are pixmaps, they differ in size bits */
      token = PeekToken(line, NULL);
      if (!token)
	break;
      /* load the file using the color reduction routines in Picture.c */
      picture = GetPicture(dpy, win, NULL, token, color_limit);
      if (!picture)
	fprintf(stderr, "%s: can't load picture %s\n", name, token);
      if (!picture)
	break;
      /* don't try to be smart with bitmaps */
      if (picture->depth != Pdepth) {
	fprintf(stderr, "%s: bitmaps not supported\n", name);
	DestroyPicture(dpy, picture);
	break;
      }
      /* copy the picture pixmap, then destroy the picture */
      /* could cache picture but it would require an extension to the struct */
      /* and I don't expect the same pixmap to be re-used very often */
      cs->width = picture->width;
      cs->height = picture->height;
      cs->pixmap = XCreatePixmap(dpy, win, cs->width, cs->height, Pdepth);
      if (picture->mask != None)
      {
	/* default to background colour where pixmap is transparent */
	XSetForeground(dpy, gc, cs->bg);
	XFillRectangle(dpy, cs->pixmap, gc, 0, 0, cs->width, cs->height);
	XSetClipMask(dpy, gc, picture->mask);
      }
      XCopyArea(dpy, picture->picture, cs->pixmap, gc, 0, 0, cs->width,
		cs->height, 0, 0);
      if (picture->mask != None)
      {
	XGCValues xgcv;
	static GC mask_gc = None;

	/* first restore the gc we modified earlier */
	XSetClipMask(dpy, gc, None);
	/* make sure we have a gc */
	if (mask_gc == None)
	{
	  /* create a gc for 1 bit depth */
	  mask_gc = XCreateGC(dpy, picture->mask, 0, &xgcv);
	}
	cs->mask = XCreatePixmap(
	  dpy, picture->mask, cs->width, cs->height, 1);
	XCopyArea(dpy, picture->mask, cs->mask, mask_gc, 0, 0, cs->width,
		  cs->height, 0, 0);
      }
      DestroyPicture(dpy, picture);
      cs->keep_aspect = (type == 2);
      cs->stretch_x = cs->stretch_y = (type != 0);
      break;
    default:
      /* test for ?Gradient */
      if (!token)
	break;
      type = toupper(token[0]);
      if (token[1])
        ret = StrEquals(&token[1], "Gradient");
      if (!ret) {
	fprintf(stderr, "%s: bad colorset pixmap specifier %s %s\n", name,
		token, line);
	break;
      }
      if (!IsGradientTypeSupported(type))
	break;

      /* create a pixmap of the gradient type */
      cs->pixmap =
	CreateGradientPixmapFromString(dpy, win, gc, type, line, &w, &h);
      cs->width = w;
      cs->height = h;
      cs->keep_aspect = False;
      cs->stretch_x = !(type == V_GRADIENT);
      cs->stretch_y = !(type == H_GRADIENT);
    }
  }

  /* calculate average background color */
  if (StrEquals(bg, average) && cs->pixmap) {
    XColor color;
    XColor *colors;
    XImage *image;
    unsigned int i, j, k = 0;
    unsigned long red = 0, blue = 0, green = 0;

    /* create an array to store all the pixmap colors in */
    colors = (XColor *)safemalloc(cs->width * cs->height * sizeof(XColor));
    /* get the pixmap into an image */
    image = XGetImage(dpy, cs->pixmap, 0, 0, cs->width, cs->height, AllPlanes,
		      ZPixmap);
    /* get each pixel */
    for (i = 0; i < cs->width; i++)
      for (j = 0; j < cs->height; j++)
	colors[k++].pixel = XGetPixel(image, i, j);
    XDestroyImage(image);
    /* look them all up, XQueryColors() can't handle more than 256 */
    for (i = 0; i < k; i += 256)
      XQueryColors(dpy, Pcmap, &colors[i], min(k - i, 256));
    /* calculate average, ignore overflow: .red is a short, red is a long */
    for (i = 0; i < k; i++) {
      red += colors[i].red;
      green += colors[i].green;
      blue += colors[i].blue;
    }
    free(colors);
    /* get it */
    XFreeColors(dpy, Pcmap, &cs->bg, 3, 0);
    k = cs->width * cs->height;
    color.red = red / k;
    color.green = green / k;
    color.blue = blue / k;
    color.flags = DoRed | DoGreen | DoBlue;
    XAllocColor(dpy, Pcmap, &color);
    cs->bg = color.pixel;
    cs->hilite = GetHilite(cs->bg);
    cs->shadow = GetShadow(cs->bg);
  }

  /* make sure the server has this to avoid races */
  XSync(dpy, False);

  /* inform fvwm of the change */
  SendText(fd, DumpColorset(n), 0);
}

/* SendToMessage options */
static char *message_options[] = {"Colorset", NULL};

static void parse_message_line(char *line) {
  char *rest;

  switch(GetTokenIndex(line, message_options, -1, &rest)) {
  case 0:
    parse_colorset(rest);
    break;
  }
}

static void parse_config(void) {
  char *line;

  /* prepare the tokenizer array, [0,1] are ImagePath and ColorLimit */
  config_options[3] = safemalloc(strlen(name) + 10);
  sprintf(config_options[3], "*%sColorset", name);

  /* set a filter on the config lines sent */
  InitGetConfigLine(fd, config_options[3]);

  /* tell fvwm what we want to receive */
  SetMessageMask(fd, M_CONFIG_INFO | M_END_CONFIG_INFO
		 | M_SENDCONFIG | M_STRING);

  /* loop on config lines, a bit strange looking because GetConfigLine
   * is a void(), have to test $line */
  while (GetConfigLine(fd, &line), line)
    parse_config_line(line);
}

static int error_handler(Display *d, XErrorEvent *e)
{
  /* Attempting to free colors or pixmaps that were not allocated by this
   * module or were never allocated at all does not cause problems */
  if (((X_FreeColors == e->request_code) && (BadAccess == e->error_code))
      || ((X_FreePixmap == e->request_code) && (BadPixmap == e->error_code)))
    return 0;

  /* other errors cause diagnostic output and stop execution */
  PrintXErrorAndCoredump(d, e, name);
  return 0;
}

static void set_signals(void) {
#ifdef HAVE_SIGACTION
  struct sigaction  sigact;

  sigemptyset(&sigact.sa_mask);
  sigaddset(&sigact.sa_mask, SIGPIPE);
  sigaddset(&sigact.sa_mask, SIGINT);
  sigaddset(&sigact.sa_mask, SIGHUP);
  sigaddset(&sigact.sa_mask, SIGTERM);
# ifdef SA_INTERRUPT
  sigact.sa_flags = SA_INTERRUPT;
# else
  sigact.sa_flags = 0;
# endif
  sigact.sa_handler = signal_handler;
  sigaction(SIGPIPE, &sigact, NULL);
  sigaction(SIGINT, &sigact, NULL);
  sigaction(SIGHUP, &sigact, NULL);
  sigaction(SIGTERM, &sigact, NULL);
#else
  /* We don't have sigaction(), so fall back to less robust methods.  */
#ifdef USE_BSD_SIGNALS
  fvwmSetSignalMask(sigmask(SIGPIPE)
		    | sigmask(SIGINT)
		    | sigmask(SIGHUP)
		    | sigmask(SIGTERM));
#endif
  signal(SIGPIPE, signal_handler);
  signal(SIGINT, signal_handler);
  signal(SIGHUP, signal_handler);
  signal(SIGTERM, signal_handler);
#ifdef HAVE_SIGINTERRUPT
  siginterrupt(SIGPIPE, 1);
  siginterrupt(SIGINT, 1);
  siginterrupt(SIGHUP, 1);
  siginterrupt(SIGTERM, 1);
#endif
#endif
}

static RETSIGTYPE signal_handler(int signal) {
  fprintf(stderr, "%s quiting on signal %d\n", name, signal);
  exit(signal);
}
